File: create.rb
class Type
  attr_reader :name, :oname
  def initialize(oname, name, funcs)
    @oname = oname
    @name = name
    @funcs = funcs
  end
  def create_typedef
    if (@name != @oname)
      return "typedef " + @oname.gsub("@", " ") + " " + @name + ";\n"
    else
      return ""
    end
  end
  def get_declare(type, typelist)
    name = @name
    if (type == "cov")
      return typelist.collect do |t|
        n = t.name
        [ "#{n} cov_#{name}_#{n}(#{name} x)", "{ return (#{n})x; }" ]
      end
    else
      return case (type)
      when "as"
        [
          [ "#{name} add_#{name}(#{name} a, #{name} b)", "{ return a + b; }" ],
          [ "#{name} sub_#{name}(#{name} a, #{name} b)", "{ return a - b; }" ],
        ]
      when "asi"
        [
          [ "#{name} add_#{name}(#{name} a, int b)", "{ return a + b; }" ],
          [ "int sub_#{name}(#{name} a, #{name} b)", "{ return a - b; }" ],
        ]
      when "md"
        [
          [ "#{name} mul_#{name}(#{name} a, #{name} b)", "{ return a * b; }" ],
          [ "#{name} div_#{name}(#{name} a, #{name} b)", "{ return a / b; }" ],
        ]
      when  "%"
        [
          [ "#{name} mod_#{name}(#{name} a, #{name} b)", "{ return a % b; }" ],
        ]
      when  "cmp"
        [
          [ "bool cmp_lt_#{name}(#{name} a, #{name} b)", "{ return a < b; }" ],
          [ "bool cmp_le_#{name}(#{name} a, #{name} b)", "{ return a <= b; }" ],
          [ "bool cmp_ge_#{name}(#{name} a, #{name} b)", "{ return a >= b; }" ],
          [ "bool cmp_gt_#{name}(#{name} a, #{name} b)", "{ return a > b; }" ],
        ]
      when "equ"
        [
          [ "bool cmp_eq_#{name}(#{name} a, #{name} b)", "{ return a == b; }" ],
          [ "bool cmp_ne_#{name}(#{name} a, #{name} b)", "{ return a != b; }" ],
        ]
      end
    end
  end
  def create_funcs(typelist)
    string = ""
    @funcs.each do |func|
      get_declare(func, typelist).each { |t| string += t[0] + " " + t[1] + "\n" }
    end
    return string
  end
  def create_declares(typelist)
    string = ""
    @funcs.each do |func|
      get_declare(func, typelist).each { |t| string += t[0] + ";\n" }
    end
    return string
  end
end

def create_include(name)
  return "#include <#{name}>\n"
end

def create_local_include(name)
  return "#include \"#{name}\"\n"
end

def create_header(types)
  string = ""
  string += create_include("stdint.h")
  string += create_include("stdbool.h")
  string += "\n"
  types.each do |type|
    string += type.create_typedef
  end
  string += "\n"
  types.each do |type|
    string += type.create_declares(types)
  end
  string += "\n"

  return string
end

def create_source(types)
  string = ""
  string += create_local_include($outputfile_inc)
  string += "\n"
  types.each do |type|
    string += type.create_funcs(types)
  end
  string += "\n"
  
  return string
end

filename = ARGV[0] || "typeinfo.txt"
$outputfile_inc = (ARGV[1] || filename) + ".h"
$outputfile_src = (ARGV[1] || filename) + ".c"
direction = ARGV[2] || "."

# Load Types

types = Array.new

File.open(filename, "r") do |file|
  while line = file.gets
    data = line.split(" ")
    next if data.empty?
    types.push(Type.new(data[0], data[1], data[2, data.size]))
  end
end

# Create

File.open(direction + "/" + $outputfile_inc, "w").print(create_header(types))
File.open(direction + "/" + $outputfile_src, "w").print(create_source(types))

File: create.py
import re
import sys

class Type:
    def __init__(self, oname, name, funclist):
        self.oname = oname
        self.name = name
        self.funclist = funclist

    def create_typedef(self):
        if (self.name != self.oname):
            return "typedef " + self.oname.replace("@", " ") + " " + self.name + ";\n"
        else:
            return ""
        
    def create_functions(self, typelist):
        return self.create_with_func(typelist, lambda t : t[0] + " " + t[1] + "\n")
        
    def create_declares(self, typelist):
        return self.create_with_func(typelist, lambda t : t[0] + ";\n")

    def create_with_func(self, typelist, f):
        string = ""
        for func in self.funclist:
            decfunclist = self.get_declare(func, typelist)
            for t in decfunclist:
                string += f(t)
        return string

    def get_format(self, format, count):
        return format % ((self.name,) * count)
        
    def get_declare(self, type, typelist):
        name = self.name
        if (type == "cov"):
            result = []
            for t in typelist:
                n = t.name
                result.append(["%s cov_%s_%s(%s x)" % (n, name, n, name), "{ return (%s)x; }" % n])
            return result
        else:
            if (type == "as"):
                return [
                    [ self.get_format("%s add_%s(%s a, %s b)", 4), "{ return a + b; }" ],
                    [ self.get_format("%s sub_%s(%s a, %s b)", 4), "{ return a - b; }" ],
                    ]
            elif (type == "asi"):
                return [
                    [ self.get_format("%s add_%s(%s a, int b)", 3), "{ return a + b; }" ],
                    [ self.get_format("int sub_%s(%s a, %s b)", 3), "{ return a - b; }" ],
                    ]
            elif (type == "md"):
                return [
                    [ self.get_format("%s mul_%s(%s a, %s b)", 4), "{ return a * b; }" ],
                    [ self.get_format("%s div_%s(%s a, %s b)", 4), "{ return a / b; }" ],
                    ]
            elif (type == '%'):
                return [
                    [ self.get_format("%s mod_%s(%s a, %s b)", 4), "{ return a % b; }" ],
                    ]
            elif (type == "cmp"):
                return [
                    [ self.get_format("bool cmp_lt_%s(%s a, %s b)", 3), "{ return a < b; }" ],
                    [ self.get_format("bool cmp_le_%s(%s a, %s b)", 3), "{ return a <= b; }" ],
                    [ self.get_format("bool cmp_ge_%s(%s a, %s b)", 3), "{ return a >= b; }" ],
                    [ self.get_format("bool cmp_gt_%s(%s a, %s b)", 3), "{ return a > b; }" ],
                ]
            elif (type == "equ"):
                return [
                    [ self.get_format("bool cmp_eq_%s(%s a, %s b)", 3), "{ return a == b; }" ],
                    [ self.get_format("bool cmp_ne_%s(%s a, %s b)", 3), "{ return a != b; }" ],
                ]
            else:
                print("Not find match type.")

def create_include(name):
    return "#include <%s>\n" % name

def create_local_include(name):
    return "#include \"%s\"\n" % name

def create_header(typelist):
    string = ""
    string += create_include("stdint.h")
    string += create_include("stdbool.h")
    string += "\n"
    for type in typelist:
        string += type.create_typedef()
    string += "\n"
    for type in typelist:
        string += type.create_declares(typelist)
    
    return string

def create_source(typelist, header_name):
    string = ""
    string += create_local_include(header_name)
    string += "\n"
    for type in typelist:
        string += type.create_functions(typelist)
    
    return string

# original_string -> identifier_string
def replace_identifier(original_string : str):
    return re.sub(r'([\:\.\%])', r'%\1', original_string)

def main():
    arglen = len(sys.argv)
    filename = sys.argv[1] if (arglen > 1) else "typeinfo.txt"
    outputfile = sys.argv[2] if (arglen > 2) else filename
    outputfile_inc = outputfile + ".h"
    outputfile_src = outputfile + ".c"
    direction = sys.argv[3] if (arglen > 3) else "."

    typelist = []

    with open(filename, "r") as file:
        for line in file.readlines():
            data = line.rstrip().split(" ")
            if (not data or data[0] == ''):
                continue
            typelist.append(Type(data[0], data[1], data[2:]))
    
    open(direction + "/" + outputfile_inc, "w").write(create_header(typelist))
    open(direction + "/" + outputfile_src, "w").write(create_source(typelist, outputfile_inc))

main()
